### PR TITLE
[RFC] Remove all occurrences of the mb_ptr2char macro

### PR DESCRIPTION
**Pull request by ZviRackover**

### CODE REVIEW

**Comment by justinmk** [comment]
Thanks for the effort. This patch was already done in https://github.com/neovim/neovim/pull/8513 (`./scripts/vim-patch -l` should not have listed it, assuming you have current Neovim master branch).

But it looks like you found more cases?

**Comment by ZviRackover** [comment]
yes, i started using the patch script and after hitting conflicts saw that it was already partially merged. Should i retitle the patch?

**Comment by justinmk** [comment]
This actually isn't the "spirit" of vim patch 8.0.0520, but more related to https://github.com/neovim/neovim/issues/7401 .

Because `mb_*` is not a function pointer at all in Nvim. 

I see you marked it WIP, but this change should be done in small steps. So if there's no known technical problem with your PR , I would mark it as RFC.

**Comment by justinmk** [comment]
@bfredl @ZyX-I Have you done something similar in existing PRs, which may conflict with this? Should we wait on this to avoid conflicts?

**Comment by ZviRackover** [comment]
I agree that #7401 is a better match. Will retitle and update commit message.

**Comment by ZviRackover** [comment]
@bfredl ok will try to remove `enc_utf8` and `has_mbyte .

**Comment by ZyX-I** [comment]
I did not do anything to remove all of some macros usages, but if I edit a code with `mb_…` I usually remove it and use `utf*` variant instead.

**Comment by ZviRackover** [comment]
This is my first pr for this project and i'm not sure what to do about the failed checks.

- QuickBuild report FAILED but the report did not mention any failed tesrs
- Some of the travis-ci jobs seemed to have failed due to environment issues

**Comment by KillTheMule** [comment]
QB: Click on "Tap report" on the left menu, then cycle through the builds by using the dropdown on the right hand side besides the "export" button. You will find some [test failures](http://neovim-qb.szakmeister.net/build/18464/tap_report/failed_tests?reportset=functionaltest-macosx-64-release). They are known and can be ignored by you.

Travis: Looks unrelated indeed, but I've never seen those before. @jamessan or @justinmk could maybe have a look.

**Comment by jamessan** [comment]
Those are strange.  I've restarted the Travis job to see if it was a temporary issue.

File: src/nvim/regexp.c [hunk 1]
```diff
@@ -6782,19 +6750,20 @@ static int vim_regsub_both(char_u *source, typval_T *expr, char_u *dest,
...
+        // Write to buffer, if copy is set.
+        if (func_one != NULL) {
           func_one = (fptr_T)(func_one(&cc, c));
-        else if (func_all != (fptr_T)NULL)
-          /* Turbo C complains without the typecast */
-          func_all = (fptr_T)(func_all(&cc, c));
-        else     /* just copy */
-          cc = c;
-
+        } else {
```
Comment (jamessan) [inline src/nvim/regexp.c]: Why was this changed into a plain `} else {` rather than kept as the `} else if (func_all != NULL) {`?

File: src/nvim/spell.c [hunk 2]
```diff
@@ -4292,23 +4289,21 @@ static void suggest_trie_walk(suginfo_T *su, langp_T *lp, char_u *fword, bool so
...
+                } else if (!soundfold && slang->sl_has_map
                            && similar_chars(slang,
-                                            mb_ptr2char(tword
-                                                        + sp->ts_twordlen
-                                                        - sp->ts_tcharlen),
-                                            mb_ptr2char(fword
-                                                        + sp->ts_fcharstart))) {
+                                            utf_ptr2char(tword + sp->ts_twordlen
+                                                         - sp->ts_tcharlen),
+                                            utf_ptr2char(fword +
```
Comment (jamessan) [inline src/nvim/spell.c]: The `+` should be on the next line if this has to be split.

File: src/nvim/spell.c [hunk 3]
```diff
@@ -4507,40 +4497,32 @@ static void suggest_trie_walk(suginfo_T *su, langp_T *lp, char_u *fword, bool so
...
+        c2 = utf_ptr2char(p + n);
       }
 
       // When the second character is NUL we can't swap.
       if (c2 == NUL) {
         PROF_STORE(sp->ts_state)
         sp->ts_state = STATE_REP_INI;
         break;
-      }
+          }
```
Comment (jamessan) [inline src/nvim/spell.c]: Indentation is wrong from here to the `#ifdef DEBUG_TRIEWALK`.

**Review by jamessan — COMMENTED** [review]
Nice work.  Just a few small things.

**Comment by janlazo** [comment]
Will break https://github.com/neovim/neovim/pull/8635 because it uses `mb_ptr2char`.

**Comment by jamessan** [comment]
> Will break #8635 because it uses mb_ptr2char.

So why don't use use `utf_ptr2char`?  `mb_ptr2char` is just a `#define`.

File: src/nvim/regexp.c (hunk 1 above)
Comment (ZyX-I) [reply src/nvim/regexp.c]: @ZviRackover Linter is completely fine with the form @jamessan suggested. You need to fix linter errors, but there is no reason to use `else { if }` form in place of `else if`, you just needed to add braces after the `if (cond)`.

Comment (ZviRackover) [reply src/nvim/regexp.c]: Done in [4f3ce52](https://github.com/neovim/neovim/pull/8665/commits/4f3ce526e644628b200257b4bcf867eb438f8c12])

File: src/nvim/spell.c (hunk 2 above)
Comment (ZviRackover) [reply src/nvim/spell.c]: Done in latest commit. Note that the ) { were move to a new line to stay within line length limits. Hope they were aligned correctly.

File: src/nvim/spell.c (hunk 3 above)
Comment (ZviRackover) [reply src/nvim/spell.c]: Done.

**Comment by ZviRackover** [comment]
The latest push should address all comments from @jamessan

File: src/nvim/spell.c [hunk 4]
```diff
@@ -4292,23 +4290,22 @@ static void suggest_trie_walk(suginfo_T *su, langp_T *lp, char_u *fword, bool so
...
-                                            mb_ptr2char(tword
-                                                        + sp->ts_twordlen
-                                                        - sp->ts_tcharlen),
-                                            mb_ptr2char(fword
-                                                        + sp->ts_fcharstart))) {
+                                            utf_ptr2char(tword + sp->ts_twordlen
+                                                         - sp->ts_tcharlen),
+                                            utf_ptr2char(fword
+                                                         + sp->ts_fcharstart))
+                           ) {
```
Comment (ZyX-I) [inline src/nvim/spell.c]: You should not move `)` on its own line. As far as I understand style guide you need to resolve the situation by either moving `!soundfold` (the very *first* part of condition) on its own line and indent it with +4 spaces or move `slang` (the very *first* argument) on its own line with +4 spaces. Following parts of condition or arguments should be aligned with the first one (like now).

File: src/nvim/spell.c [hunk 5]
```diff
@@ -4292,23 +4290,22 @@ static void suggest_trie_walk(suginfo_T *su, langp_T *lp, char_u *fword, bool so
                     && utf_iscomposing(utf_ptr2char(fword
                                                     + sp->ts_fcharstart))) {
                   sp->ts_score -= SCORE_SUBST - SCORE_SUBCOMP;
-                } else if (!soundfold
-                           && slang->sl_has_map
+                } else if (!soundfold && slang->sl_has_map
```
Comment (ZyX-I) [inline src/nvim/spell.c]: Do not join lines like this, lengthy multipart condition is more readable when each part starts on its own line.

File: src/nvim/spell.c (hunk 4 above)
Comment (ZviRackover) [reply src/nvim/spell.c]: Ok, i'll go for moving `!soundfold` to a new line. If all fails, there is a possibility of refactoring out the expression `tword + sp->ts_twordlen - sp->ts_tcharlen` as it repeats in neighboring conditions

Comment (ZviRackover) [reply src/nvim/spell.c]: Here's what can be done following what you siggest @ZyX-I. Only problem is that the rightmost brace reaches column 81 and the linter complains only about that.
Is there another possiblity for fitting this code?
```
                } else if (!soundfold
                           && slang->sl_has_map
                           && similar_chars(slang,
                                            utf_ptr2char(tword
                                                         + sp->ts_twordlen
                                                         - sp->ts_tcharlen),
                                            utf_ptr2char(fword
                                                         + sp->ts_fcharstart))) {
                  // For a similar character adjust score from
                  // SCORE_SUBST to SCORE_SIMILAR.
                  sp->ts_score -= SCORE_SUBST - SCORE_SIMILAR;
                }
```

Comment (ZyX-I) [reply src/nvim/spell.c]: @ZviRackover It is not what I suggested and I do not see `!soundfold` on a new line there. What I suggested is

```C
                } else if (
                    !soundfold
                    && slang->sl_has_map
                    && similar_chars(slang,
                                     utf_ptr2char(tword
                                                  + sp->ts_twordlen
                                                  - sp->ts_tcharlen),
                                     utf_ptr2char(fword
                                                  + sp->ts_fcharstart))) {
                  // For a similar character adjust score from
                  // SCORE_SUBST to SCORE_SIMILAR.
                  sp->ts_score -= SCORE_SUBST - SCORE_SIMILAR;
                }
```

Same thing may be done with `slang` as well, I think doing both would be cleaner:

```C
                } else if (
                    !soundfold
                    && slang->sl_has_map
                    && similar_chars(
                        slang,
                        utf_ptr2char(tword + sp->ts_twordlen - sp->ts_tcharlen),
                        utf_ptr2char(fword + sp->ts_fcharstart))) {
                  // For a similar character adjust score from
                  // SCORE_SUBST to SCORE_SIMILAR.
                  sp->ts_score -= SCORE_SUBST - SCORE_SIMILAR;
                }
```

Comment (ZviRackover) [reply src/nvim/spell.c]: Thanks. Now i see what you mean.

**Comment by ZviRackover** [comment]
Latest push should address all comments by @ZyX-I

**Comment by ZviRackover** [comment]
If there are no additional comments, would it be possible to merge this pr? My concern is that the changes to 25 files in hundreds of lines  are in risk of conflict with other patches. Thanks

File: src/nvim/spell.c (hunk 3 above)
Comment (oni-link) [reply src/nvim/spell.c]: The new change shits the code to the right, the old indentation was okay.

File: src/nvim/spell.c [hunk 6]
```diff
@@ -3648,7 +3645,8 @@ static void suggest_trie_walk(suginfo_T *su, langp_T *lp, char_u *fword, bool so
   bool goodword_ends;
 #ifdef DEBUG_TRIEWALK
   // Stores the name of the change made at each level.
-  char_u changename[MAXWLEN][80];
+  static const size_t changenameElemSize = 80;
```
Comment (oni-link) [inline src/nvim/spell.c]: What is the advantage of the additional `static` here?

Comment (ZviRackover) [reply src/nvim/spell.c]: It narrowa the scope of the variable to current file.

Comment (oni-link) [reply src/nvim/spell.c]: It narrows it down to function scope and keeps the value between function calls. But why do we want this instead of just a constant value? Is it more efficient?

Comment (ZyX-I) [reply src/nvim/spell.c]: @ZviRackover It does not, not in this context. Neither it does “narrow down to function scope”, all variables declared inside function already have only this scope, no matter `static` or not. Just `static` variables declared in a function are essentially *global* variables; you would not get a symbol for them so you can’t just access that outside of function, but you *do* get a constant address which may be returned from the function, one which will continue to be valid after function was finished. This is not needed.

File: src/nvim/spell.c [hunk 7]
```diff
@@ -3648,7 +3645,8 @@ static void suggest_trie_walk(suginfo_T *su, langp_T *lp, char_u *fword, bool so
...
+  char_u changename[MAXWLEN][changenameElemSize];
```
Comment (ZyX-I) [inline src/nvim/spell.c]: The result looks much like VLA. I am not sure what exactly compiler would do, but you must not define dimensions using anything of which compiler not thinks like it is “an integer constant expression”. According to the standard this construct defenitely *is* a VLA:

> An integer constant expression99) shall have integer type and shall only have operands that are integer constants, enumeration constants, character constants, sizeof expressions whose results are integer constants, and floating constants that are the immediate operands of casts. Cast operators in an integer constant expression shall only convert arithmetic types to integer types, except as part of an operand to the sizeof operator.

though compiler may think otherwise.

Also note that `changenameElemSize` is *not* a valid name for variable according to the style guide. For constants (though I never saw it used for constants scoped to function) it is `kChangenameElemSize`, for variables it is `changename_elem_size`.

And you ultimately not need any variable here at all. I have checked how `void main(){char a[10][20]; printf("%zu", sizeof(a[0]));}` program works and it returns 20 on gcc, clang and tcc, so you should use `sizeof` in place of variable. Also note that comment that is now before `changenameElemSize` definition actually applies to `changename` and *not* `changenameElemSize` which is incorrect.

Comment (ZviRackover) [reply src/nvim/spell.c]: I totally missed that this is in a function, so of course the `static` is not needed. Good catch.

**Comment by ZviRackover** [comment]
Latest revision removes the redundant `static` discussed above.

**Comment by ZviRackover** [comment]
The latest revision removes `changenameElemSize`

File: src/nvim/spell.c (hunk 3 above)
Comment (ZviRackover) [reply src/nvim/spell.c]: Done

File: src/nvim/spell.c [hunk 8]
```diff
@@ -4542,9 +4525,10 @@ static void suggest_trie_walk(suginfo_T *su, langp_T *lp, char_u *fword, bool so
       if (c2 != NUL && TRY_DEEPER(su, stack, depth, SCORE_SWAP)) {
         go_deeper(stack, depth, SCORE_SWAP);
 #ifdef DEBUG_TRIEWALK
-        sprintf(changename[depth], "%.*s-%s: swap %c and %c",
-            sp->ts_twordlen, tword, fword + sp->ts_fidx,
-            c, c2);
+        snprintf(changename[depth], sizeof changename[0],
```
Comment (ZyX-I) [inline src/nvim/spell.c]: Missing parenthesis: while `sizeof` being operator can be used without them it is always written like if it was a function.

Comment (ZviRackover) [reply src/nvim/spell.c]: Done. 
Didn't find this rule in the relevant style guide section, https://neovim.io/develop/style-guide.xml#sizeof. If it is indeed not listed, should it be added?

File: src/nvim/regexp.c [hunk 9]
```diff
@@ -1107,13 +1103,9 @@ static int get_coll_element(char_u **pp)
   char_u      *p = *pp;
 
   if (p[0] != NUL && p[1] == '.') {
-    if (has_mbyte)
-      l = (*mb_ptr2len)(p + 2);
+    l = (*mb_ptr2len)(p + 2);
```
Comment (bfredl) [inline src/nvim/regexp.c]: `utfc_ptr2len`

Comment (ZviRackover) [reply src/nvim/regexp.c]: Done

**Comment by ZviRackover** [comment]
@justinmk @ZyX-I do you have any additional comments? I see the milestones for this pr are bumped without it being pushed and would like to understand if i am expected to do anything to make progress.

File: src/nvim/charset.c [hunk 10]
```diff
@@ -1375,7 +1375,7 @@ void getvvcol(win_T *wp, pos_T *pos, colnr_T *start, colnr_T *cursor,
     ptr = ml_get_buf(wp->w_buffer, pos->lnum, false);
 
     if (pos->col < (colnr_T)STRLEN(ptr)) {
-      int c = (*mb_ptr2char)(ptr + pos->col);
+      int c = utf_ptr2char(ptr + pos->col);
```
Comment (janlazo) [inline src/nvim/charset.c]: Add `const` unless this changes in a future vim-patch.

File: src/nvim/edit.c [hunk 11]
```diff
@@ -2311,24 +2311,14 @@ static void ins_compl_longest_match(compl_T *match)
...
-        c1 = mb_ptr2char(p);
-        c2 = mb_ptr2char(s);
-      } else {
-        c1 = *p;
-        c2 = *s;
-      }
-      if (match->cp_icase ? (mb_tolower(c1) != mb_tolower(c2))
-          : (c1 != c2)) {
+      c1 = utf_ptr2char(p);
+      c2 = utf_ptr2char(s);
```
Comment (janlazo) [inline src/nvim/edit.c]: `c1`, and `c2` are `const` now?

File: src/nvim/edit.c [hunk 12]
```diff
@@ -8494,7 +8482,7 @@ int ins_copychar(linenr_T lnum)
   if ((colnr_T)temp > curwin->w_virtcol)
     ptr = prev_ptr;
 
-  c = (*mb_ptr2char)(ptr);
+  c = utf_ptr2char(ptr);
```
Comment (janlazo) [inline src/nvim/edit.c]: Is this `const int` now?

File: src/nvim/ex_docmd.c [hunk 13]
```diff
@@ -2900,7 +2900,7 @@ const char * set_one_cmd_context(
     p = (const char *)xp->xp_pattern;
     while (*p != NUL) {
       if (has_mbyte) {
-        c = mb_ptr2char((const char_u *)p);
+        c = utf_ptr2char((const char_u *)p);
```
Comment (janlazo) [inline src/nvim/ex_docmd.c]: If you are keeping only the branch where `has_mbyte` is true, can you be consistent with the rest?

File: src/nvim/mbyte.c [hunk 14]
```diff
@@ -674,7 +674,7 @@ int mb_ptr2char_adv(const char_u **const pp)
 {
   int c;
 
-  c = (*mb_ptr2char)(*pp);
+  c = utf_ptr2char(*pp);
```
Comment (janlazo) [inline src/nvim/mbyte.c]: `const int c`

File: src/nvim/mbyte.c [hunk 15]
```diff
@@ -687,7 +687,7 @@ int mb_cptr2char_adv(const char_u **pp)
...
```
Comment (janlazo) [inline src/nvim/mbyte.c]: `const int c`

File: src/nvim/ex_docmd.c [hunk 16]
```diff
@@ -2922,7 +2922,7 @@ const char * set_one_cmd_context(
         len = 0;          /* avoid getting stuck when space is in 'isfname' */
         while (*p != NUL) {
           if (has_mbyte) {
-            c = mb_ptr2char((const char_u *)p);
+            c = utf_ptr2char((const char_u *)p);
```
Comment (janlazo) [inline src/nvim/ex_docmd.c]: See https://github.com/neovim/neovim/pull/8665/files#r207964834

File: src/nvim/ex_getln.c [hunk 17]
```diff
@@ -3329,8 +3329,9 @@ static bool cmdline_paste(int regname, bool literally, bool remcr)
       for (w = ccline.cmdbuff + ccline.cmdpos; w > ccline.cmdbuff; ) {
         if (has_mbyte) {
           len = (*mb_head_off)(ccline.cmdbuff, w - 1) + 1;
-          if (!vim_iswordc(mb_ptr2char(w - len)))
+          if (!vim_iswordc(utf_ptr2char(w - len))) {
```
Comment (janlazo) [inline src/nvim/ex_getln.c]: See https://github.com/neovim/neovim/pull/8665/files#r207964834

File: src/nvim/ex_getln.c [hunk 18]
```diff
@@ -3841,13 +3842,13 @@ ExpandOne (
     for (len = 0; xp->xp_files[0][len]; len += mb_len) {
       if (has_mbyte) {
         mb_len = (* mb_ptr2len)(&xp->xp_files[0][len]);
-        c0 = (* mb_ptr2char)(&xp->xp_files[0][len]);
+        c0 = utf_ptr2char(&xp->xp_files[0][len]);
```
Comment (janlazo) [inline src/nvim/ex_getln.c]: See https://github.com/neovim/neovim/pull/8665/files#r207964834

File: src/nvim/ex_getln.c [hunk 19]
```diff
@@ -3841,13 +3842,13 @@ ExpandOne (
...
       } else {
         c0 = xp->xp_files[0][len];
       }
       for (i = 1; i < xp->xp_numfiles; ++i) {
         if (has_mbyte) {
-          ci =(* mb_ptr2char)(&xp->xp_files[i][len]);
+          ci = utf_ptr2char(&xp->xp_files[i][len]);
```
Comment (janlazo) [inline src/nvim/ex_getln.c]: See https://github.com/neovim/neovim/pull/8665/files#r207964834

File: src/nvim/fold.c [hunk 20]
```diff
@@ -1787,8 +1787,9 @@ char_u *get_foldtext(win_T *wp, linenr_T lnum, linenr_T lnume,
         int len;
 
         if (has_mbyte && (len = (*mb_ptr2len)(p)) > 1) {
-          if (!vim_isprintc((*mb_ptr2char)(p)))
+          if (!vim_isprintc(utf_ptr2char(p))) {
```
Comment (janlazo) [inline src/nvim/fold.c]: See https://github.com/neovim/neovim/pull/8665/files#r207964834

File: src/nvim/message.c [hunk 21]
```diff
@@ -2870,8 +2870,9 @@ do_dialog (
       retval = 1;
       for (i = 0; hotkeys[i]; ++i) {
         if (has_mbyte) {
-          if ((*mb_ptr2char)(hotkeys + i) == c)
+          if (utf_ptr2char(hotkeys + i) == c) {
```
Comment (janlazo) [inline src/nvim/message.c]: See https://github.com/neovim/neovim/pull/8665/files#r207964834

File: src/nvim/message.c [hunk 22]
```diff
@@ -2912,7 +2913,7 @@ copy_char (
 
   if (has_mbyte) {
     if (lowercase) {
-      c = mb_tolower((*mb_ptr2char)(from));
+      c = mb_tolower(utf_ptr2char(from));
```
Comment (janlazo) [inline src/nvim/message.c]: See https://github.com/neovim/neovim/pull/8665/files#r207964834

File: src/nvim/ops.c [hunk 23]
```diff
@@ -3634,10 +3634,10 @@ int do_join(size_t count,
       if (has_mbyte) {
         cend = curr + currsize;
         MB_PTR_BACK(curr, cend);
-        endcurr1 = (*mb_ptr2char)(cend);
+        endcurr1 = utf_ptr2char(cend);
```
Comment (janlazo) [inline src/nvim/ops.c]: See https://github.com/neovim/neovim/pull/8665/files#r207964834

File: src/nvim/option.c [hunk 24]
```diff
@@ -6853,27 +6856,27 @@ void find_mps_values(int *initc, int *findc, int *backwards, int switchit)
     if (has_mbyte) {
       char_u *prev;
 
-      if (mb_ptr2char(ptr) == *initc) {
+      if (utf_ptr2char(ptr) == *initc) {
```
Comment (janlazo) [inline src/nvim/option.c]: See https://github.com/neovim/neovim/pull/8665/files#r207964834

File: src/nvim/regexp.c [hunk 25]
```diff
@@ -5474,7 +5442,7 @@ regrepeat (
       } else if (rex.reg_line_lbr && *scan == '\n' && WITH_NL(OP(p))) {
         scan++;
       } else if (has_mbyte && (len = (*mb_ptr2len)(scan)) > 1) {
-        if ((cstrchr(opnd, (*mb_ptr2char)(scan)) == NULL) == testval) {
+        if ((cstrchr(opnd, utf_ptr2char(scan)) == NULL) == testval) {
```
Comment (janlazo) [inline src/nvim/regexp.c]: See https://github.com/neovim/neovim/pull/8665/files#r207964834

File: src/nvim/spell.c [hunk 26]
```diff
@@ -4684,7 +4648,7 @@ static void suggest_trie_walk(suginfo_T *su, langp_T *lp, char_u *fword, bool so
         p = fword + sp->ts_fidx;
         if (has_mbyte) {
           n = MB_CPTR2LEN(p);
-          c = mb_ptr2char(p);
+          c = utf_ptr2char(p);
```
Comment (janlazo) [inline src/nvim/spell.c]: See https://github.com/neovim/neovim/pull/8665/files#r207964834

File: src/nvim/regexp.c [hunk 27]
```diff
@@ -6878,10 +6846,7 @@ static int vim_regsub_both(char_u *source, typval_T *expr, char_u *dest,
                 }
                 dst += 2;
               } else {
-                if (has_mbyte)
-                  c = mb_ptr2char(s);
-                else
-                  c = *s;
+                c = utf_ptr2char(s);
```
Comment (janlazo) [inline src/nvim/regexp.c]: See https://github.com/neovim/neovim/pull/8665/files#r207964834

File: src/nvim/spell.c [hunk 28]
```diff
@@ -2609,16 +2611,11 @@ static bool spell_iswordp(char_u *p, win_T *wp)
...
 {
-  int c;
-
-  if (has_mbyte) {
-    c = mb_ptr2char(p);
-    if (c > 255) {
-      return spell_mb_isword_class(mb_get_class(p), wp);
-    }
-    return spelltab.st_isw[c];
+  int c = utf_ptr2char(p);
```
Comment (janlazo) [inline src/nvim/spell.c]: `const int c`

File: src/nvim/regexp_nfa.c [hunk 29]
```diff
@@ -5033,13 +5033,8 @@ static int nfa_regmatch(nfa_regprog_T *prog, nfa_state_T *start,
...
 
-    if (has_mbyte) {
-      curc = (*mb_ptr2char)(reginput);
-      clen = (*mb_ptr2len)(reginput);
-    } else {
-      curc = *reginput;
-      clen = 1;
-    }
+    curc = utf_ptr2char(reginput);
+    clen = utfc_ptr2len(reginput);
```
Comment (janlazo) [inline src/nvim/regexp_nfa.c]: You can declare and initialize `curc` on same line. Same applies for `clen`.

File: src/nvim/regexp.c [hunk 30]
```diff
@@ -3468,10 +3446,7 @@ static long bt_regexec_both(char_u *line,
   if (prog->regmust != NULL) {
     int c;
 
-    if (has_mbyte)
-      c = (*mb_ptr2char)(prog->regmust);
-    else
-      c = *prog->regmust;
+    c = utf_ptr2char(prog->regmust);
```
Comment (janlazo) [inline src/nvim/regexp.c]: `int c = ...`

File: src/nvim/charset.c (hunk 10 above)
Comment (ZviRackover) [reply src/nvim/charset.c]: I didn't see any coding style guideline in the wiki requiring `const` on variables which are not pointers.
I don't have a strong objection to adding `const` in cases such as this, but i don't see much benefit in it.

File: src/nvim/ex_docmd.c (hunk 13 above)
Comment (ZviRackover) [reply src/nvim/ex_docmd.c]: Good catch, thanks. Will fix all these cases you pointed out.

File: src/nvim/charset.c (hunk 10 above)
Comment (bfredl) [reply src/nvim/charset.c]: Especially with a small local block there is really not much benefit to add const to a local variable binding (it just trivially repeats info already visible in the block), we should focus the effort to pointer targets (where const makes it a distinct type), and possibly bindings in really large scopes.

File: src/nvim/misc1.c [hunk 31]
```diff
@@ -1749,11 +1748,7 @@ del_lines (
 
 int gchar_pos(pos_T *pos)
 {
-  char_u      *ptr = ml_get_pos(pos);
-
-  if (has_mbyte)
-    return (*mb_ptr2char)(ptr);
-  return (int)*ptr;
+  return utf_ptr2char(ml_get_pos(pos));
```
Comment (justinmk) [inline src/nvim/misc1.c]: Note for next time: it's often good to have a local (`ptr`) anyway, for debugging.

**Review by justinmk — APPROVED** [review]
